### PR TITLE
Added cancellation support to `TileProvider` and surrounding mechanisms

### LINKED ISSUE #1430: [FEATURE] Cancel unnecessary tile requests
### What is the bug?

Hello!

I use example WEB app project with openstreetmap as tiles provider (home page) and map work properly, but when i change openstreetmap to mapbox map work very slow.

Main difference between providers - OSM use HTTP 2 and mapBox use HTTP 1.1 (with their pipeline). When i am scrolling on mapBox - map is setting many requests to queue and don't interrupt of them, it load hundreds of useless images when map move before show target place.

How i can interrupt request which out of view?

Thanks!

### What is the expected behaviour?

Map will interrupt useless requests

### How can we reproduce this issue?

_No response_

### Do you have a potential solution?

_No response_

### Can you provide any other information?

_No response_

### Platforms Affected

Web

### Severity

Minimum: Allows normal functioning

### Frequency

Consistently: Always occurs at the same time and location

### Requirements

- [X] I agree to follow this project's [Code of Conduct](https://github.com/fleaflet/flutter_map/blob/master/CODE_OF_CONDUCT.md)
- [X] My Flutter/Dart installation is unaltered, and `flutter doctor` finds no relevant issues
- [X] I am using the [latest stable version](https://pub.dev/packages/flutter_map) of this package
- [X] I have checked the FAQs section on the documentation website
- [X] I have checked for similar issues which may be duplicates

### PR DESCRIPTION
**Pull request by JaffaKetchup**
Closes #1430. Note that this doesn't provide an implementation of a `TileProvider` that actually cancels HTTP requests, as this is currently unavailable without Dio (see #1430's thread for more info). It just implements the infrastructure necessary to allow an implementation.

To test the cancellation support, install Dio, and try these code blocks on top of this branch. Then, run on the web, and open the browser's DevTools Network tab. Observe how some requests are marked as "(cancelled)". Note that this is more apparent with slower tile servers, so using the default OpenStreetMap tile server likely won't cause too many cancellations.

<details>
<summary>network_tile_provider.dart</summary>

```dart
import 'package:flutter/rendering.dart';
import 'package:flutter_map/src/layer/tile_layer/tile_coordinates.dart';
import 'package:flutter_map/src/layer/tile_layer/tile_layer.dart';
import 'package:flutter_map/src/layer/tile_layer/tile_provider/base_tile_provider.dart';
import 'package:flutter_map/src/layer/tile_layer/tile_provider/network_image_provider.dart';
import 'package:http/http.dart';
import 'package:http/retry.dart';

/// [TileProvider] to fetch tiles from the network
///
/// By default, a [RetryClient] is used to retry failed requests. 'dart:http'
/// or 'dart:io' might be needed to override this.
///
/// On the web, the 'User-Agent' header cannot be changed as specified in
/// [TileLayer.tileProvider]'s documentation, due to a Dart/browser limitation.
class NetworkTileProvider extends TileProvider {
  /// [TileProvider] to fetch tiles from the network
  ///
  /// By default, a [RetryClient] is used to retry failed requests. 'dart:http'
  /// or 'dart:io' might be needed to override this.
  ///
  /// On the web, the 'User-Agent' header cannot be changed as specified in
  /// [TileLayer.tileProvider]'s documentation, due to a Dart/browser limitation.
  NetworkTileProvider({
    super.headers,
    BaseClient? httpClient,
  }) : httpClient = httpClient ?? RetryClient(Client());

  /// The HTTP client used to make network requests for tiles
  final BaseClient httpClient;

  @override
  bool get supportsCancelLoading => true;

  @override
  ImageProvider getImageWithCancelLoadingSupport(
    TileCoordinates coordinates,
    TileLayer options,
    Future<void> cancelLoading,
  ) =>
      FlutterMapNetworkImageProvider(
        url: getTileUrl(coordinates, options),
        fallbackUrl: getTileFallbackUrl(coordinates, options),
        headers: headers,
        httpClient: httpClient,
        cancelLoading: cancelLoading,
      );
}
```

</details>

<details>
<summary>network_image_provider.dart</summary>

```dart
import 'dart:async';
import 'dart:ui';

import 'package:dio/dio.dart';
import 'package:flutter/foundation.dart';
import 'package:flutter/painting.dart';
import 'package:flutter_map/src/layer/tile_layer/tile_provider/base_tile_provider.dart';
import 'package:http/http.dart';

/// Dedicated [ImageProvider] to fetch tiles from the network
@immutable
class FlutterMapNetworkImageProvider
    extends ImageProvider<FlutterMapNetworkImageProvider> {
  /// The URL to fetch the tile from (GET request)
  final String url;

  /// The URL to fetch the tile from (GET request), in the event the original
  /// [url] request fails
  final String? fallbackUrl;

  /// The HTTP client to use to make network requests
  final BaseClient httpClient;

  /// The headers to include with the tile fetch request
  final Map<String, String> headers;

  final Future<void> cancelLoading;

  /// Dedicated [ImageProvider] to fetch tiles from the network
  const FlutterMapNetworkImageProvider({
    required this.url,
    required this.fallbackUrl,
    required this.headers,
    required this.httpClient,
    required this.cancelLoading,
  });

  @override
  ImageStreamCompleter loadImage(
    FlutterMapNetworkImageProvider key,
    ImageDecoderCallback decode,
  ) {
    final chunkEvents = StreamController<ImageChunkEvent>();

    return MultiFrameImageStreamCompleter(
      codec: _loadAsync(key, chunkEvents, decode),
      chunkEvents: chunkEvents.stream,
      scale: 1,
      debugLabel: url,
      informationCollector: () => [
        DiagnosticsProperty('URL', url),
        DiagnosticsProperty('Fallback URL', fallbackUrl),
        DiagnosticsProperty('Current provider', key),
      ],
    );
  }

  @override
  Future<FlutterMapNetworkImageProvider> obtainKey(
    ImageConfiguration configuration,
  ) =>
      SynchronousFuture<FlutterMapNetworkImageProvider>(this);

  Future<Codec> _loadAsync(
    FlutterMapNetworkImageProvider key,
    StreamController<ImageChunkEvent> chunkEvents,
    ImageDecoderCallback decode, {
    bool useFallback = false,
  }) async {
    final cancelToken = CancelToken();
    cancelLoading.then((_) => cancelToken.cancel());

    final Uint8List bytes;
    try {
      final dio = Dio();
      final response = await dio.get<Uint8List>(
        useFallback ? fallbackUrl ?? '' : url,
        cancelToken: cancelToken,
        options: Options(
          headers: headers,
          responseType: ResponseType.bytes,
        ),
      );
      bytes = response.data!;
    } on DioException catch (err) {
      if (CancelToken.isCancel(err)) {
        return decode(
          await ImmutableBuffer.fromUint8List(TileProvider.transparentImage),
        );
      }
      if (useFallback || fallbackUrl == null) rethrow;
      return _loadAsync(key, chunkEvents, decode, useFallback: true);
    } catch (_) {
      if (useFallback || fallbackUrl == null) rethrow;
      return _loadAsync(key, chunkEvents, decode, useFallback: true);
    }

    return decode(await ImmutableBuffer.fromUint8List(bytes));
  }
}
```

</details>

---

Also involves a heavy (and breaking for a minority of implementations) cleanup of the `TileProvider` interface, to reduce the number of private methods, reduce the amount of code that must be overridden to achieve a custom behaviour, and reduce the scope of the `TileLayer`.

### CODE REVIEW

**Comment by josxha** [comment]
Hi @JaffaKetchup, thanks for implementing this awesome feature!
I created a demo app, throttled the internet speed in the chrome dev tools and did some testing on web.
image

The cancellation works flawlessly.
image

I haven't looked at the complete code yet but noticed two things:

File: lib/src/layer/tile_layer/tile_image_manager.dart [hunk 1]
```diff
@@ -125,10 +125,16 @@ class TileImageManager {
...
+      tile.imageProvider = layer.tileProvider.supportsCancelLoading
+          ? layer.tileProvider.getImageWithCancelLoadingSupport(
+              tileBounds.atZoom(tile.coordinates.z).wrap(tile.coordinates),
+              layer,
+              tile.cancelLoading.future,
+            )
+          : layer.tileProvider.getImage(
+              tileBounds.atZoom(tile.coordinates.z).wrap(tile.coordinates),
+              layer,
+            );
```
Comment (josxha) [inline lib/src/layer/tile_layer/tile_image_manager.dart:137]: Why is this logic not inside the TileProvider if the tile manager asks the TileProvider what function to call anyways.

File: lib/src/layer/tile_layer/tile_provider/base_tile_provider.dart [hunk 2]
```diff
@@ -1,84 +1,284 @@
...
+  ImageProvider getImageWithCancelLoadingSupport(
+    TileCoordinates coordinates,
+    TileLayer options,
+    Future<void> cancelLoading,
+  ) {
+    throw UnimplementedError(
+      'A `TileProvider` that overrides `supportsCancelLoading` to `true` must '
+      'override `getImageWithCancelLoadingSupport`',
+    );
+  }
```
Comment (josxha) [inline lib/src/layer/tile_layer/tile_provider/base_tile_provider.dart:97]: Wouldn't it be possible to merge `.getImage()` and `getImageWithCancelLoadingSupport()` into one function? The implementation will probably only override one function anyways and could handle additional logic inside the function.
Your solution is probably best for backwards compatibility but it feels a bit odd to me.

Comment (JaffaKetchup) [reply lib/src/layer/tile_layer/tile_provider/base_tile_provider.dart:97]: Yeah, I was aiming for backwards compatibility. I have made breaking changes in this PR, but most implementations won't be affected.
If the other maintainers think that this isn't necessary, I'm happy to simplify it.

File: lib/src/layer/tile_layer/tile_image_manager.dart (hunk 1 above)
Comment (JaffaKetchup) [reply lib/src/layer/tile_layer/tile_image_manager.dart:137]: See below, in summary, to reduce breaking chnages.

**Review by TesteurManiak — APPROVED** [review]
LGTM, maybe wwe should add your code snippets as examples in the test app to showcase the cancellation mechanism 🤔

**Comment by JaffaKetchup** [comment]
@TesteurManiak That's a good idea, I'll do that. Someone else said they might make a package, but it's easy enough just to make an example for. Just to double check, you're happy with the breaking refactoring and non-breaking cancellation support of the `TileProvider`?

**Comment by TesteurManiak** [comment]
> you're happy with the breaking refactoring and non-breaking cancellation support of the TileProvider?

Yeah, deprecation messages are clear and the changes are not too intrusive 👍

**Pull request merged**